### PR TITLE
SI-7502 removing non-existent element from ListMap leaves it unchaged.

### PR DESCRIPTION
**Pull request by vigdorchik**
Current imperative version constructs a new ListMap regardless of
the fact the map doesn't contain the element. Replace it with the
tail-recursive variant that conserves. Also replace some usages with
the invariant now held.

JIRA: https://issues.scala-lang.org/browse/SI-7502

### CODE REVIEW

**Comment by adriaanm** [comment]
review by @JamesIry

**Comment by paulp** [comment]
LGTM, but what I would like to know is, what if anything documents what the caller can do and what the callee can expect when a method is called on an immutable collection, if the result of the method is of the same type as the receiver?

The really important case from my point of view is "filter". There are a bunch of immutable collections which do not return themselves even when filter does nothing. But in the absence of any clear guidance, you never know whether someone has code which depends on 'a ne a.filter(_ => true)'.

```
scala> val x = (1 to 10).toList
x: List[Int] = List(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)

scala> val y = x filter (_ => true)
y: List[Int] = List(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)

scala> x eq y
res1: Boolean = false
```

**Comment by vigdorchik** [comment]
For immutable maps checking `(m - k) ne m` provides O(1) way to check that the binding for `k` is present in `m`.
For other methods, such as `filter` I don't see such simple invariants that justify additional complexity in the implementation. It's always possible to check with O(n) for things like: did I filter anything?

**Comment by vigdorchik** [comment]
Ah, and maybe scaladoc for immutable maps should reflect this invariant more visibly.

**Comment by paulp** [comment]
I'm not sure what you mean about additional complexity; I'm asking because filter and other methods which often return themselves are woefully inefficient for no good reason, and I would like to fix them. But it would be easy for there to be code out there which depends on the current woefully inefficient behavior, and there is no documentation which says "don't depend on this."

File: src/library/scala/collection/immutable/ListMap.scala [hunk 1]
```diff
@@ -189,32 +189,17 @@ extends AbstractMap[A, B]
...
-        val elem = lst.head
-        val stbl = acc
-        acc = new stbl.Node(elem._1, elem._2)
-        lst = lst.tail
-      }
-      acc
-    }
-
+    override def - (k: A): ListMap[A, B1] = remove0(k, this, Nil)
+
```
Comment (JamesIry) [inline src/library/scala/collection/immutable/ListMap.scala:193]: I slightly prefer having 'remove0' as a method defined inside the '-' method

**Pull request merged**